amethyst_tiles/src/morton/mod.rs

```diff
@@ -0,0 +1,231 @@
+/// of Morton encoding is used, considered extremely fast but still slightly slower than BMI2 intrinsics.
+#[derive(Default, Clone)]
+pub struct Encoder;
+impl CoordinateEncoder for Encoder {
+    fn from_dimensions(_: u32, _: u32, _: u32) -> Self {
+        Self {}
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        Some(encode(x, y, z))
+    }
+    #[inline]
+    fn decode(&self, morton: u32) -> Option<(u32, u32, u32)> {
+        Some(decode(morton))
+    }
+}
+
+/// 3D Morton (Z-Order) encoding implementation.
+/// This implementation uses the `bmi2` CPU intrinsic if it is available via the `bitintr` crate. If this instruction
+/// set is not available, it falls back on simpler computation methods. Using these CPU instruction optimizations requires
+/// `RUSTFLAGS=-C target-feature=+bmi2`. If this target feature is not provided, a LUT (Look Up Table) implementation
+/// of Morton encoding is used, considered extremely fast but still slightly slower than BMI2 intrinsics.
+///
+/// This implementation only performs 2D morton encoding on any given Z-level, while providing Z-levels ia a standard
+/// flat-array multiplicative manner. This means that each Z-level is contiguous in memory, but its inner coordinates
+/// are still Z-order encoded for some spatial locality.
+#[derive(Default, Clone)]
+pub struct Encoder2D {
+    dimensions: (u32, u32, u32),
+    len: u32,
+}
+impl CoordinateEncoder for Encoder2D {
+    fn from_dimensions(x: u32, y: u32, z: u32) -> Self {
+        //#[cfg(debug_assertions)]
+        //{
+        //    assert_eq!(x % 16, 0);
+        //    assert_eq!(x % 16, 0);
+        //}
+
+        Self {
+            dimensions: (x, y, z),
+            len: x * y,
+        }
+    }
+
+    #[inline]
+    fn encode(&self, x: u32, y: u32, z: u32) -> Option<u32> {
+        use bitintr::Pdep;
+
+        #[cfg(debug_assertions)]
+        {
+            let check = u32::max_value() / 3;
+            if x > check || y > check || z > check {
+                return None;
+            }
+        }
+
+        let morton = (x.pdep(0x5555_5555) | y.pdep(0xAAAA_AAAA)) + (z * self.len);
+
+        Some(morton)
+    }
+    #[inline]
+    fn decode(&self, mut morton: u32) -> Option<(u32, u32, u32)> {
+        use bitintr::Pext;
+
+        let z = morton / self.len;
+        morton -= z * self.len;
+
+        Some((morton.pext(0x5555_5555), morton.pext(0xAAAA_AAAA), z))
+    }
+}
+
+#[inline]
+#[cfg(not(target_feature = "bmi2"))]
+pub fn encode(x: u32, y: u32, z: u32) -> u32 {
+    morton_encode_lut(x, y, z)
+}
+
+#[inline]
+#[cfg(not(target_feature = "bmi2"))]
+pub fn decode(morton: u32) -> (u32, u32, u32) {
+    morton_decode_lut(morton)
+}
+
+#[inline]
+pub fn min(morton1: u32, morton2: u32) -> u32 {
+    let d1 = decode(morton1);
+    let d2 = decode(morton2);
+
+    if d1.0 > d2.0 || d1.1 > d2.1 || d1.2 > d2.2 {
```

Comment:
Is this the logic you want for sure? This logic is constructed in such a way that `min` is not commutative, or even idempotent (think what happens when x is equal). I believe you wanted to chain the comparators instead of running them in parallel.

Really what you want is to implement an comparator and use it in both `min` and `max` to select the value.

```rust
#[inline]
pub fn cmp(morton1: u32, morton2: u32) -> Ordering {
	let (x1, y1, z1) = decode(morton1);
	let (x2, y2, z2) = decode(morton2);
	z1.cmp(z2).then(|| y1.cmp(y2)).then(|| x1.cmp(x2))
}
```

Or, if the (z, y, x) order is not necessary and (x, y, z) is good enough, then we can use the `Ord` on tuples (it chains the comparators from left to right).

```rust
#[inline]
pub fn cmp(morton1: u32, morton2: u32) -> Ordering {
	decode(morton1).cmp(decode(morton2))
}
```

Reply:
> I believe you wanted to chain the comparators instead of running them in parallel.

rust is lazy-evaluated - that means the comparisons run linearly, left to right, and fail after the first failure and are chained. I'm not sure what you are talking about here needing to use .then comparisons here.